In [3]:
import pandas as pd
import numpy as npt
import os
import warnings
warnings.filterwarnings('ignore')


### IMPORT DATA

In [4]:
df_lainnya_20 = pd.read_csv('./fix_dataset/dataset_InflasiPendidikan/Inflasi Pendidikan 2020-Lainnya.csv')
df_lainnya_21 = pd.read_csv('./fix_dataset/dataset_InflasiPendidikan/Inflasi Pendidikan 2021-Lainnya.csv')
df_lainnya_22 = pd.read_csv('./fix_dataset/dataset_InflasiPendidikan/Inflasi Pendidikan 2022-Lainnya.csv')
df_lainnya_23 = pd.read_csv('./fix_dataset/dataset_InflasiPendidikan/Inflasi Pendidikan 2023-Lainnya.csv')

### DATA CLEANING

* Merge Data
* Check Null, Unique, Object, etc from Data

In [5]:
df_lainnya_20.shape

(91, 13)

In [6]:
temp1 = pd.merge(df_lainnya_20, df_lainnya_21, on='Kota')
temp2 = pd.merge(df_lainnya_22, df_lainnya_23, on='Kota')
df_lainnya = pd.merge(temp1, temp2, on='Kota')
df_lainnya.head()

Kota  01/01/2020  01/02/2020  01/03/2020  01/04/2020  \
0          KOTA MEULABOH        0.00         0.0         0.0         0.0   
1        KOTA BANDA ACEH        0.00         0.0         0.0         0.0   
2       KOTA LHOKSEUMAWE        0.00         0.0         0.0         0.0   
3           KOTA SIBOLGA       10.52         0.0         0.0         0.0   
4  KOTA PEMATANG SIANTAR        0.00         0.0         0.0         0.0   

   01/05/2020  01/06/2020  01/07/2020  01/08/2020  01/09/2020  ...  \
0         0.0         0.0         0.0         0.0        0.00  ...   
1         0.0         0.0         0.0         0.0        0.83  ...   
2         0.0         0.0         0.0         0.0        1.87  ...   
3         0.0         0.0         0.0         0.0        0.00  ...   
4         0.0         0.0         0.0         0.0        0.00  ...   

   01/03/2023  01/04/2023  01/05/2023  01/06/2023  01/07/2023  01/08/2023  \
0        1.41        3.43         0.0         0.0         0.0         0.0   
1        0.00        0.00         0.0         0.0         0.0         0.0   
2        0.00        0.00         0.0         0.0         0.0         0.0   
3        0.00        0.00         0.0         0.0         0.0         0.0   
4        0.00        0.00         0.0         0.0         0.0         0.0   

   01/09/2023  01/10/2023  01/11/2023  01/12/2023  
0         0.0         0.0         NaN         NaN  
1         0.0         0.0         NaN         NaN  
2         0.0         0.0         NaN         NaN  
3         0.0         0.0         NaN         NaN  
4         0.0         0.0         NaN         NaN  

[5 rows x 49 columns]

In [7]:
temp = df_lainnya.melt(id_vars=['Kota'], var_name='Date', value_name='InflationRate')
temp['Date'] = pd.to_datetime(temp['Date'], format='%d/%m/%Y')
temp['Year'] = temp['Date'].dt.year
temp['Month'] = temp['Date'].dt.month
temp.drop('Date', axis=1, inplace=True)
df_lainnya = temp[['Kota', 'Year', 'Month', 'InflationRate']]

In [8]:
df_lainnya['Kota'].unique()

array(['KOTA MEULABOH', 'KOTA BANDA ACEH', 'KOTA LHOKSEUMAWE',
       'KOTA SIBOLGA', 'KOTA PEMATANG SIANTAR', 'KOTA MEDAN',
       'KOTA PADANGSIDIMPUAN', 'KOTA GUNUNGSITOLI', 'KOTA PADANG',
       'KOTA BUKITTINGGI', 'TEMBILAHAN', 'KOTA PEKANBARU', 'KOTA DUMAI',
       'BUNGO', 'KOTA JAMBI', 'KOTA PALEMBANG', 'KOTA LUBUKLINGGAU',
       'KOTA BENGKULU', 'KOTA BANDAR LAMPUNG', 'KOTA METRO',
       'TANJUNG PANDAN', 'KOTA PANGKAL PINANG', 'KOTA BATAM',
       'KOTA TANJUNG PINANG', 'DKI JAKARTA', 'KOTA BOGOR',
       'KOTA SUKABUMI', 'KOTA BANDUNG', 'KOTA CIREBON', 'KOTA BEKASI',
       'KOTA DEPOK', 'KOTA TASIKMALAYA', 'CILACAP', 'PURWOKERTO', 'KUDUS',
       'KOTA SURAKARTA', 'KOTA SEMARANG', 'KOTA TEGAL', 'KOTA YOGYAKARTA',
       'JEMBER', 'BANYUWANGI', 'SUMENEP', 'KOTA KEDIRI', 'KOTA MALANG',
       'KOTA PROBOLINGGO', 'KOTA MADIUN', 'KOTA SURABAYA',
       'KOTA TANGERANG', 'KOTA CILEGON', 'KOTA SERANG', 'SINGARAJA',
       'KOTA DENPASAR', 'KOTA MATARAM', 'KOTA BIMA', 'WAINGAPU'

In [9]:
df_lainnya.isnull().sum()

Kota               0
Year               0
Month              0
InflationRate    182
dtype: int64

In [10]:
df_lainnya['InflationRate'].fillna(0, inplace=True)

In [11]:
df_lainnya.isnull().sum()

Kota             0
Year             0
Month            0
InflationRate    0
dtype: int64

In [12]:
df_lainnya["Category"] = "Pendidikan"
df_lainnya.head()

Kota  Year  Month  InflationRate    Category
0          KOTA MEULABOH  2020      1           0.00  Pendidikan
1        KOTA BANDA ACEH  2020      1           0.00  Pendidikan
2       KOTA LHOKSEUMAWE  2020      1           0.00  Pendidikan
3           KOTA SIBOLGA  2020      1          10.52  Pendidikan
4  KOTA PEMATANG SIANTAR  2020      1           0.00  Pendidikan

### PRE-PROCESSING

* Divide Label and Features
* Splitting Data
* Standard Scaler

In [13]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [14]:
labelEncoder = LabelEncoder()
df_lainnya['Kota'] = labelEncoder.fit_transform(df_lainnya['Kota'])
df_lainnya['Category'] = labelEncoder.fit_transform(df_lainnya['Category'])

In [15]:
X = df_lainnya.drop('InflationRate', axis=1)
y = df_lainnya['InflationRate']

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [17]:
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

### DATA MODELLING

* Model = (Support Vector Regression) SVR
* Model = Random Forest Regression

In [18]:
from sklearn.svm import SVR
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression

In [19]:
model_svr = SVR(kernel='linear')
model_rfr = RandomForestRegressor(n_estimators=10, random_state=42)
model_lr = LinearRegression()
model_svr.fit(X_train, y_train)
model_rfr.fit(X_train, y_train)
model_lr.fit(X_train, y_train)

LinearRegression()

### CHECKING RESULT

* Make Y Predict
* Checking MSE, MAE, R2 Score

In [20]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [21]:
Y_pred_svr = model_svr.predict(X_train)
Y_pred_rfr = model_rfr.predict(X_train)
Y_pred_lr = model_lr.predict(X_train)

In [22]:
#SVR
print('SVR')
print('Mean Squared Error:', mean_squared_error(y_train, Y_pred_svr))
print('Mean Absolute Error:', mean_absolute_error(y_train, Y_pred_svr))
print('R^2 Score:', r2_score(y_train, Y_pred_svr))

SVR
Mean Squared Error: 1.1259027314446424
Mean Absolute Error: 0.272627102396531
R^2 Score: -0.005242591137297525


In [23]:
#Random Forest Regressor
print('Random Forest Regressor')
print('Mean Squared Error:', mean_squared_error(y_train, Y_pred_rfr))
print('Mean Absolute Error:', mean_absolute_error(y_train, Y_pred_rfr))
print('R^2 Score:', r2_score(y_train, Y_pred_rfr))


Random Forest Regressor
Mean Squared Error: 0.21427576244991411
Mean Absolute Error: 0.13180709788208356
R^2 Score: 0.8086876275833403


In [24]:
#Linear Regression
print('Linear Regression')
print('Mean Squared Error:', mean_squared_error(y_train, Y_pred_lr))
print('Mean Absolute Error:', mean_absolute_error(y_train, Y_pred_lr))
print('R^2 Score:', r2_score(y_train, Y_pred_lr))

Linear Regression
Mean Squared Error: 1.1190185858853625
Mean Absolute Error: 0.3354296518428979
R^2 Score: 0.0009037980013959501


Tentu, berikut adalah kesimpulan untuk hasil evaluasi yang diberikan:

1. **Mean Squared Error (MSE):**
   - Random Forest Regressor memiliki MSE terendah (0.2143), menunjukkan tingkat kesalahan prediksi yang lebih rendah dibandingkan dengan model lainnya.
   - Linear Regression memiliki MSE tertinggi (1.1190), menunjukkan tingkat kesalahan prediksi yang lebih tinggi.

2. **Mean Absolute Error (MAE):**
   - Random Forest Regressor memiliki MAE terendah (0.1318), menunjukkan tingkat kesalahan prediksi yang lebih rendah dibandingkan dengan model lainnya.
   - Linear Regression memiliki MAE tertinggi (0.3354), menunjukkan tingkat kesalahan prediksi yang lebih tinggi.

3. **R^2 Score:**
   - Random Forest Regressor memiliki R^2 Score tertinggi (0.8087), menunjukkan tingkat eksplanasi varian yang lebih baik oleh model.
   - Linear Regression memiliki R^2 Score terendah (-0.0052), menunjukkan bahwa model ini tidak mampu menjelaskan variasi yang signifikan dalam data.

Dengan demikian, berdasarkan evaluasi tersebut, Random Forest Regressor juga merupakan pilihan yang lebih baik untuk dataset ini, dengan kinerja prediksi yang lebih baik dibandingkan dengan model Linear Regression dan SVR.

### SAVE MODEL FOR : DF_LAINNYA

In [26]:
import joblib
joblib.dump(model_rfr, 'model_lainnya.pkl')

['model_lainnya.pkl']